src/brokers/surrealdb/broker.rs

```
@@ -0,0 +1,475 @@
+                    queued_message,
+                    "Could not try consume (get actual message)",
+                )
+                .await;
+                match msg {
+                    Ok(msg) => Ok(Some(msg)),
+                    Err(e) => Err(e),
+                }
+            }
+            None => Ok(None), //////// CRITICAL AREA ENDS (None branch) //////
+        }
+    }
+
+    /// Consumes a message from the specified queue, blocking until a message is available.
+    /// Uses live querying, so if there are no messages yet, it will block efficiently without polling
+    /// # Arguments
+    /// * `queue_name` - The name of the queue.
+    ///
+    /// # Returns
+    /// A `Result` containing the message as a `String`, or a `BroccoliError` on failure.
+    async fn consume(
+        &self,
+        queue_name: &str,
+        options: Option<ConsumeOptions>,
+    ) -> Result<InternalBrokerMessage, BroccoliError> {
+        // first of all, we try to consume without blocking, and return if we have messages
+        let resp = Self::try_consume(&self, queue_name, options).await?;
+        if let Some(message) = resp {
+            return Ok(message);
+        }
+
+        // if there were no messages, we block using a live query and wait
+        let db = self.check_connected()?;
+        let queue_table = utils::queue_table(queue_name);
+        let mut stream = db
+            .select(queue_table)
+            .range(
+                vec![Value::default(), Value::default()] // note default is 'None'
+                ..vec![Value::from_str("time::now()").unwrap_or(Value::default()),Value::default()],
+            ) // should notify when future becomes present
+            .live()
+            .await
+            .map_err(|err| BroccoliError::Broker(format!("Could not consume: {:?}", err)))?;
+        let mut queued_message: Result<InternalSurrealDBBrokerQueuedMessageRecord, BroccoliError> =
+            Err(BroccoliError::NotImplemented);
```

I dont think this is the right error to use here

---

The idea there is that TTL is not implemented (nor easily implementable) in surrealdb so doing a fail fast, would a `Publish` be a better choice?

---

Is this functionality for TTL? I thought it was for consuming messages

---

Not sure if I understand the question, I will rephrase.

Thea idea is that if we detect that the user has requested the TTL feature in the publish options we fail fast and return that the feature is not implemented in surrealdb. 
It's not easy to implement as surrealdb has not concept of timers, there are some ideas on how to do it but they are best left for the future.

I can remove the check and let the user publish, but it could be misleading as the user would expect messages to autoexpire (and they won't).

Could add a clarifying comment